### ISSUE TITLE
Pre-Draft: add Uniswap to on-chain loans

### ISSUE THREAD

**Issue by DunnCreativeSS** [Stakedllc/YIPs#21]
Why
Uniswap, as I undertand it, is a liquidity pool I can put funds into so that market makers and other parties can execute trades at leverage or for 'free-ish.' These funds then produce yield as the market makers have paid fees which gets added to the pot, then available for the liquidity providers to later withdraw. I may be incorrect in my understanding or I may have forgotten something.

Solution
https://github.com/Uniswap

Considerations
I'm certainly hoping their contracts are open and allow these types of on-chain interactions from other smart contracts.

image

https://etherscan.io/address/0x09cabec1ead1c0ba254b09efb3ee13841712be14#writeContract

Uniswap DAI (and everybody else born from this factory - 1000s exchanges) do indeed have 'addLiquidity' 'removeLiquidity' onchain :)

Additional Info
This appears to be something that would allow for yield :) Could RAY diversify and balance tokens thru exchanging them for Ether on the Uniswap exchanges should RAY put a new % into a new basket of top-performing Uniswap exchanges? This would also add fees to the pool, but cost some gas. I'm unsure how RAY balances things in on-chain loans, is it possible to have that kind of behavior (rebalancing without new gas) on Uniswap? @dpurhar27

Refer to https://github.com/Stakedllc/robo-advisor-yield/issues/6 for existing conversation.

**Comment by dpurhar27** [comment]
As mentioned, please refer to https://github.com/Stakedllc/robo-advisor-yield/issues/6 for some problems to work through for this suggestion.

**Comment by DunnCreativeSS** [comment]
Hey @dpurhar27 I think you'd answered yourself:

'It's probable in the future RAY will let users opt-in to opportunities which aren't fully passive, such as this, and then we could offer it without a hedging solution.'

While we might be legally required to only use on-chain solutions, it limits the amount of publicly available fully on-chain opportunities there are - if we have opt-ins for extra ways to achieve yield that might have some additional dangers (in this case impermanent loss).

We can link them off-site to a number of places to DYOR, and also supply them with nifty visuals like the one on that Medium:

![image](https://user-images.githubusercontent.com/22843601/74177626-81daa880-4c10-11ea-9a2d-34f28ce6501e.png)

In reality, people want to achieve a few things:

1. Gain passive yield
2. Risk more to gain more, if informed
3. Diversify in each 1. 2.
4. Diversify between 1's and 2's baskets, according to my risk aversion. When my factory friends set up their retirement funds, the fund manager says 'well we can do 100% safe (certain bonds, GICs in Canada) but you won't make much. We can do 33 33 33% safe medium and super high risk, but you might lose 33-50% of that.. so be prepared. Or, we can do the 100% super high risk and you really only might make your money back if you're lucky or you might be a millionaire in less than a year - this one's a gamble. Risk only what you can afford to lose.' Notice the gambling terms? That's where we'd win - because on paper and realistically we've tried our best to educate and inform but people are greedy and egotistical and usually (in crypto) have an appetite for risk. If we feed them, we get a % back. Let's feed them.
5. Have all of this happen behind-the-scenes
6. Look at pretty graphs. Have you seen how fulcrum and probably a bunch of others update your interest earned in real-time down to like 10 decimal places of a cent? It's beautiful. We should incorporate that.

**Comment by dpurhar27** [comment]
Yes, as mentioned earlier, I think riskier opportunities makes sense in time.

re: updating interest to extreme precision, we currently don't have it because we believe it adds no value. I think some love it, but having that many decimals complicates the GUI, and is only useful for tiny deposits, etc. However, if someone wanted to build their own RAY GUI, then naturally they're free to design it anyway they wish. We've had a few third-party's do this, including [DefiSnap](https://www.defisnap.io/#/dashboard) and Math Wallet, and we encourage it!
Some surveying and A/B testing of target customers would need to be done with positive results for us to change our GUI.
